### PR TITLE
Fix internal HubConnection state when handshake fails

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 1]
```diff
@@ -132,24 +132,25 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
                 // Start the connection
                 var connection = _connectionFactory();
                 await connection.StartAsync(_protocol.TransferFormat);
-                _connectionState = new ConnectionState(connection, this);
+                var connectionState = new ConnectionState(connection, this);
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: Maybe call this `startingConnectionState` to make it more obvious that it isn't ready to use yet?

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 2]
```diff
@@ -132,24 +132,25 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
...
                 {
                     Log.ErrorStartingConnection(_logger, ex);
 
                     // Can't have any invocations to cancel, we're in the lock.
-                    await _connectionState.Connection.DisposeAsync();
+                    await connectionState.Connection.DisposeAsync();
                     throw;
                 }
 
+                _connectionState = connectionState;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: Add a comment why we wait until here before setting `_connectionState`.

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 3]
```diff
@@ -494,15 +495,15 @@ private void CheckDisposed()
             }
         }
 
-        private async Task HandshakeAsync()
+        private async Task HandshakeAsync(ConnectionState connectionState)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: Also rename to startingConnectionState?

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 4]
```diff
@@ -132,24 +132,26 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
                 // Start the connection
                 var connection = _connectionFactory();
                 await connection.StartAsync(_protocol.TransferFormat);
-                _connectionState = new ConnectionState(connection, this);
+                var startingConnectionState = new ConnectionState(connection, this);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: nit: initialConnectionState?

**Pull request merged**